### PR TITLE
[Frontend] Use GenerateName instead of Name in resource templates

### LINKED ISSUE #166: [Frontend] Use GenerateName instead of Name in resource templates
# Problem
Duplicate auto-completed resource names on new.
(I don't know whether it is called `auto-complete`.)

e.g.)
- Create `New Pod`. (the pod name will be `pod1`.)
- Create `New Pod`.(the pod name will be `pod2`.)
- Delete `pod1`.
- Try to create `New Pod`. (the name will be `pod2`) ← `pod2` already exists.

### PR DESCRIPTION
**Pull request by aibarbetta**
#### What type of PR is this?
/kind bug

#### What this PR does / why we need it:
Use metadata.generateName instead of metadata.name with ordinal suffix. Avoid name collisions when resources are created, deleted and created again, e.g.:

* New Pod + Apply -> (the pod name will be `pod1`)
* New Pod + Apply  -> (the pod name will be `pod2`)
* Delete `pod1`
* New Pod + Apply (the name will be `pod2`) <- `pod2` already exists.

#### Which issue(s) this PR fixes:

Fixes #166 

#### Special notes for your reviewer:

/label tide/merge-method-squash

### CODE REVIEW

File: web/api/v1/pod.ts [hunk 1]
```diff
@@ -3,19 +3,18 @@ import { k8sInstance, namespaceURL } from "@/api/v1/index";
...
       delete req.metadata.managedFields;
     }
-    const res = await k8sInstance.patch<V1Pod>(
-      namespaceURL +
-        `/pods/${req.metadata.name}?fieldManager=simulator&force=true`,
+    const res = await k8sInstance.post<V1Pod>(
+      namespaceURL + `/pods?fieldManager=simulator&force=true`,
       req,
-      { headers: { "Content-Type": "application/apply-patch+yaml" } }
+      { headers: { "Content-Type": "application/yaml" } }
```
[the remark this answers is not in the document]
Comment (aibarbetta) [reply web/api/v1/pod.ts:17]: Correct, I tried using `generateName`  with apply but the API rejected it with this `cannot use generate name with apply`. I'll create a new `createPod` function and revert my changes in `applyPod`. Thanks!

File: web/api/v1/node.ts [hunk 2]
```diff
@@ -3,6 +3,26 @@ import { AxiosInstance } from "axios";
 
 export default function nodeAPI(k8sInstance: AxiosInstance) {
   return {
+    createNode: async (req: V1Node) => {
```
Comment (sanposhiho) [inline web/api/v1/node.ts:6]:
```suggestion
    // createNode accepts only Node that has .metadata.GeneratedName. 
    // If you want to create a Node that has .metadata.Name, use applyNode instead.
    createNode: async (req: V1Node) => {
```

File: web/store/node.ts [hunk 3]
```diff
@@ -65,7 +65,15 @@ export default function nodeStore() {
     },
 
     async apply(n: V1Node) {
-      await nodeAPI.applyNode(n);
+      if (n.metadata?.name) {
+        await nodeAPI.applyNode(n);
+      } else if (!n.metadata?.name && n.metadata?.generateName) {
```
Comment (sanposhiho) [inline web/store/node.ts:70]:
```suggestion
      } else if (!n.metadata?.name && n.metadata?.generateName) {
      // This Node can be expected to be a newly created Node. So, use `createNode` instead.
```

File: web/store/node.ts [hunk 4]
```diff
@@ -65,7 +65,15 @@ export default function nodeStore() {
...
+        await nodeAPI.createNode(n);
+      } else {
+        throw new Error(`
+        failed to apply node: node has no metadata.name or metadata.generateName
```
Comment (sanposhiho) [inline web/store/node.ts:74]:
```suggestion
        failed to apply node: node should have metadata.name or metadata.generateName
```

**Review by sanposhiho — COMMENTED** [review]
👍 
Small nits. Please apply the reviews for other resources as well.

File: web/store/pod.ts [hunk 5]
```diff
@@ -88,7 +88,15 @@ export default function podStore() {
...
     async apply(p: V1Pod) {
-      await podAPI.applyPod(p);
+      if (p.metadata?.name) {
+        await podAPI.applyPod(p);
+      } else if (!p.metadata?.name && p.metadata?.generateName) {
+        await podAPI.createPod(p);
+      } else {
+        throw new Error(`
+        failed to apply pod: pod has no metadata.name or metadata.generateName
+        `);
```
Comment (196Ikuchil) [inline web/store/pod.ts:98]:
```suggestion
        throw new Error("failed to apply pod: pod has no metadata.name or metadata.generateName");
```
Is there any reason you would use "`"?

File: web/store/node.ts (hunk 3 above)
Comment (196Ikuchil) [inline web/store/node.ts:70]:
```suggestion
      if (n.metadata?.name) {
        await nodeAPI.applyNode(n);
      } else if (n.metadata?.generateName) {
```
This condition (`!n.metadata?.name`) will always be true?

File: web/api/v1/pod.ts [hunk 6]
```diff
@@ -4,6 +4,26 @@ import { AxiosInstance } from "axios";
 
 export default function podAPI(k8sInstance: AxiosInstance) {
   return {
+    createPod: async (req: V1Pod) => {
+      try {
+        if (!req.metadata?.generateName) {
+          throw new Error(`metadata.generateName is not provided`);
```
Comment (196Ikuchil) [inline web/api/v1/pod.ts:10]:
```suggestion
          throw new Error("metadata.generateName is not provided");
```

File: web/api/v1/pod.ts [hunk 7]
```diff
@@ -4,6 +4,26 @@ import { AxiosInstance } from "axios";
...
+        }
+        req.kind = "Pod";
+        req.apiVersion = "v1";
+        if (req.metadata.managedFields) {
+          delete req.metadata.managedFields;
+        }
+        const res = await k8sInstance.post<V1Pod>(
+          namespaceURL + `/pods?fieldManager=simulator&force=true`,
```
Comment (196Ikuchil) [inline web/api/v1/pod.ts:18]:
```suggestion
          namespaceURL + "/pods?fieldManager=simulator&force=true",
```
Essentially, I think `"` should be used in other existing string values...

**Review by 196Ikuchil — COMMENTED** [review]
Hi! Thank you for your PR. I left some comments✋

File: web/store/node.ts (hunk 3 above)
Comment (aibarbetta) [reply web/store/node.ts:70]: it will, I left it for clarity but I can remove it since it is redundant

**Comment by aibarbetta** [comment]
Thank you for the reviews @sanposhiho and @196Ikuchil, I just pushed the fixes, let me know if I missed anything

**Comment by 196Ikuchil** [comment]
Thank you!
/lgtm

**Review by sanposhiho — APPROVED** [review]
スクリーンショット 2022-05-29 13 51 36

Nice work @aibarbetta. I confirmed it works well.
Thanks and welcome to our project.

/approve

**Pull request merged**